--- a/src/hirobot_tof_plane_seg.cpp
+++ b/src/hirobot_tof_plane_seg.cpp
@@ -6,7 +6,7 @@
 #include <pcl/segmentation/sac_segmentation.h>
 #include <pcl/segmentation/extract_polygonal_prism_data.h>
 #include <pcl/surface/concave_hull.h>
-#include <pcl/visualization/pcl_visualizer.h>
+//#include <pcl/visualization/pcl_visualizer.h>
 #include "pcl_conversions/pcl_conversions.h"
 
 #include <pcl/ModelCoefficients.h>
@@ -21,7 +21,7 @@
 #include<pcl/io/pcd_io.h>
 #include<pcl/point_types.h>
 #include <pcl/common/transforms.h>
-#include <pcl/visualization/pcl_visualizer.h>
+//#include <pcl/visualization/pcl_visualizer.h>
 
 
 #define TRANSFORMATION_TYPE_X 0
