From 46f4baa6808c17aa9bf2695174116adcf762c2a7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Andreas=20M=C3=BCller?= <schnitzeltony@googlemail.com>
Date: Wed, 11 Nov 2015 22:02:54 +0100
Subject: [PATCH 4/5] fix location(s) for our cmake config
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Upstream-Status: Inappropriate [configuration]

Signed-off-by: Andreas Müller <schnitzeltony@googlemail.com>
---
 src/core/CMakeLists.txt | 8 +-------
 1 file changed, 1 insertion(+), 7 deletions(-)

diff --git a/src/core/CMakeLists.txt b/src/core/CMakeLists.txt
index 5e333a9..ec31803 100644
--- a/src/core/CMakeLists.txt
+++ b/src/core/CMakeLists.txt
@@ -252,7 +252,7 @@ file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/CLuceneConfig.cmake" "
     set(CLUCENE_SOVERSION ${CLUCENE_SOVERSION})
 ")
 install(FILES "${CMAKE_CURRENT_BINARY_DIR}/CLuceneConfig.cmake"
-        DESTINATION ${LIB_DESTINATION}/CLuceneConfig.cmake)
+        DESTINATION ${LIB_DESTINATION}/cmake)
 
 # install pkg-config file
 IF(NOT WIN32)
@@ -267,16 +267,10 @@ IF ( LUCENE_SYS_INCLUDES )
     install(FILES ${clucene-shared_BINARY_DIR}/CLucene/clucene-config.h 
             DESTINATION ${LUCENE_SYS_INCLUDES}/CLucene  
             COMPONENT development)
-    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/CLuceneConfig.cmake 
-            DESTINATION ${LUCENE_SYS_INCLUDES}/CLucene  
-            COMPONENT development)
 ELSE ( LUCENE_SYS_INCLUDES )
     install(FILES ${clucene-shared_BINARY_DIR}/CLucene/clucene-config.h 
             DESTINATION include/CLucene
             COMPONENT development)
-    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/CLuceneConfig.cmake 
-            DESTINATION include/CLucene
-            COMPONENT development)
 ENDIF ( LUCENE_SYS_INCLUDES )
 
 
-- 
2.1.0

