From 37dff3758407e74cc092008a8a2aeacc22c7e25d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Andreas=20M=C3=BCller?= <schnitzeltony@gmail.com>
Date: Wed, 21 Sep 2022 22:59:46 +0200
Subject: [PATCH] Fix build with glibc 2.36
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Andreas Müller <schnitzeltony@gmail.com>
---
 src/core/CLucene/document/DateTools.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/core/CLucene/document/DateTools.cpp b/src/core/CLucene/document/DateTools.cpp
index f99ab25..46e6fa1 100644
--- a/src/core/CLucene/document/DateTools.cpp
+++ b/src/core/CLucene/document/DateTools.cpp
@@ -8,6 +8,7 @@
 
 #include "DateTools.h"
 #include "CLucene/util/Misc.h"
+#include <time.h>
 
 CL_NS_USE(util)
 CL_NS_DEF(document)
-- 
2.35.3

