diff -uprN a/libelf/gelf.h b/libelf/gelf.h
--- a/libelf/gelf.h	2024-04-28 16:16:17.442820402 +0800
+++ b/libelf/gelf.h	2024-04-28 16:14:45.835813283 +0800
@@ -83,7 +83,9 @@ typedef Elf64_Rel GElf_Rel;
 typedef Elf64_Rela GElf_Rela;
 
 /* Relative relocation entry (in section of type SHT_RELR).  */
-typedef Elf64_Relr GElf_Relr;
+#ifndef SHT_RELR
+  typedef Elf64_Relr GElf_Relr;
+#endif
 
 /* Program segment header.  */
 typedef Elf64_Phdr GElf_Phdr;
