--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -13,6 +13,7 @@ find_package(mipi_camera REQUIRED)
 find_package(camera REQUIRED)
 find_package(zxing REQUIRED)
 find_package(barcode_interface REQUIRED)
+find_package(std_msgs REQUIRED)
 
 # uncomment the following section in order to fill in
 # further dependencies manually.
@@ -22,10 +23,8 @@ set(AI_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/../../../../)
 set(ZXING_ROOT ${AI_ROOT}/sample/zxing/src/zxing-cpp-1.4.0/core/src/)
 set(COMMON_ROOT ${AI_ROOT}/component/common/)
 set(BARCODE_ROOT ${AI_ROOT}/component/barcode_detection/)
-set(SDK_ROOT ${AI_ROOT}/../../3rdparty/platform/ss928/org/smp/a55_linux/)
 link_directories(
-	${SDK_ROOT}/mpp/out/lib
-        ${SDK_ROOT}/mpp/out/lib/svp_npu
+    ${CMAKE_SYSROOT}/usr/lib/svp_npu
 	${AI_ROOT}/sample/mipi_camera/src/ffmpeglib/lib
 )
 
--- a/package.xml
+++ b/package.xml
@@ -15,6 +15,7 @@
   <depend>camera</depend>
   <depend>zxing</depend>
   <depend>barcode_interface</depend>
+  <depend>std_msgs</depend>
   <export>
     <build_type>ament_cmake</build_type>
   </export>
