--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -13,7 +13,6 @@ find_package(ament_cmake REQUIRED)
 # find_package(<dependency> REQUIRED)
 set(SRC_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/src)
 set(AI_ROOT  ${CMAKE_CURRENT_SOURCE_DIR}/../../)
-set(SDK_ROOT ${AI_ROOT}/../../3rdparty/platform/ss928/org/smp/a55_linux/)
 
 add_compile_options(-fvisibility=hidden)
 
@@ -71,7 +70,6 @@ target_link_libraries(camera
 target_include_directories(camera PUBLIC
         $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include/camera>
 	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/common>
-	$<BUILD_INTERFACE:${SDK_ROOT}/mpp/out/include>
 	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/ffmpeglib/include>
 	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/ffmpeglib/include/libavformat>
 	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/src/ffmpeglib/include/libavcodec>
