--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -19,12 +19,10 @@ find_package(mipi_camera REQUIRED)
 find_package(tf2_geometry_msgs REQUIRED)
 
 set(AI_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/../../../../)
-set(SDK_ROOT ${AI_ROOT}/../../3rdparty/platform/ss928/org/smp/a55_linux/)
 set(PDFD_ROOT ${AI_ROOT}/component/body_detection/)
 
 link_directories(
-		${SDK_ROOT}/mpp/out/lib
-		${SDK_ROOT}/mpp/out/lib/svp_npu
+		${CMAKE_SYSROOT}/usr/lib/svp_npu
 		${AI_ROOT}/sample/mipi_camera/src/ffmpeglib/lib
 		${PDFD_ROOT}/lib
 )
@@ -36,7 +34,6 @@ add_executable(pose_srv_node
 		${CMAKE_CURRENT_SOURCE_DIR}/../../../common/ivp_draw.cpp)
 
 target_include_directories(pose_srv_node PUBLIC
-		$<BUILD_INTERFACE:${SDK_ROOT}/mpp/out/include>
 		$<BUILD_INTERFACE:${AI_ROOT}/sample/camera/include>
 		$<BUILD_INTERFACE:${AI_ROOT}/sample/mipi_camera/include>
 		$<BUILD_INTERFACE:${AI_ROOT}/component/common>
