diff -uprN a/tests/TfLiteYoloV3Big-Armnn/TfLiteYoloV3Big-Armnn.cpp b/tests/TfLiteYoloV3Big-Armnn/TfLiteYoloV3Big-Armnn.cpp
--- a/tests/TfLiteYoloV3Big-Armnn/TfLiteYoloV3Big-Armnn.cpp	2024-07-17 07:58:32.244108640 +0000
+++ b/tests/TfLiteYoloV3Big-Armnn/TfLiteYoloV3Big-Armnn.cpp	2024-07-17 07:56:31.949756463 +0000
@@ -21,6 +21,7 @@
 #include <fstream>
 #include <iostream>
 #include <cmath>
+#include <iterator>
 
 using namespace armnnTfLiteParser;
 using namespace armnn;
