diff -urpN a/delegate/cmake/Modules/FindTfLite.cmake b/delegate/cmake/Modules/FindTfLite.cmake
--- a/delegate/cmake/Modules/FindTfLite.cmake	2024-07-21 21:41:44.597112645 +0800
+++ b/delegate/cmake/Modules/FindTfLite.cmake	2024-07-21 21:40:25.159199380 +0800
@@ -22,10 +22,10 @@ find_library(TfLite_LIB NAMES "libtensor
 # If the static library was found, gather all of its dependencies
 if (TfLite_LIB MATCHES .a$)
     message("-- Static tensorflow lite library found, using for ArmNN build")
-    find_library(TfLite_abseilstrings_LIB "libabsl_strings.a"
-                 PATH ${TFLITE_LIB_ROOT}/_deps/abseil-cpp-build/absl/strings)
-    find_library(TfLite_abseil_synchronization_LIB "libabsl_synchronization.a"
-                 PATH ${TFLITE_LIB_ROOT}/_deps/abseil-cpp-build/absl/synchronization)
+    find_library(TfLite_abseilstrings_LIB "libabsl_strings.so"
+                 PATH ${TFLITE_LIB_ROOT})
+    find_library(TfLite_abseil_synchronization_LIB "libabsl_synchronization.so"
+                 PATH ${TFLITE_LIB_ROOT})
 
 
     find_library(TfLite_farmhash_LIB "libfarmhash.a"
