diff --git a/arch/arm64/boot/compressed/Makefile b/arch/arm64/boot/compressed/Makefile
index 6ab779cbe..2618f5c19 100644
--- a/arch/arm64/boot/compressed/Makefile
+++ b/arch/arm64/boot/compressed/Makefile
@@ -66,7 +66,9 @@ ifeq ($(CONFIG_CPU_ENDIAN_BE8),y)
 LDFLAGS_vmlinux += --be8
 endif
 # ?
+ifeq ($(CONFIG_CC_IS_GCC), y)
 LDFLAGS_vmlinux += -p
+endif
 # Report unresolved symbol references
 LDFLAGS_vmlinux += --no-undefined
 # Delete all temporary local symbols
