diff --git a/arch/arm64/boot/compressed/head.S b/arch/arm64/boot/compressed/head.S
index ff3c43db7..7e4d1e2bf 100644
--- a/arch/arm64/boot/compressed/head.S
+++ b/arch/arm64/boot/compressed/head.S
@@ -219,7 +219,8 @@ restart:
          * the relocation code added. This avoids overwriting
          * ourself when the offset is small.
          */
-        add     x10, x10, #((reloc_code_end - restart + 0x1000) & ~0xFFF)
+        mov     x13, #((reloc_code_end - restart + 0x1000) & ~0xFFF)
+        add     x10, x10, x13
         bic	x10, x10, #0xFFF
 
         /* Get start of code we want to copy and align it down. */
