diff --git a/Lib/sysconfig.py b/Lib/sysconfig.py
index 046447f..5123040 100644
--- a/Lib/sysconfig.py
+++ b/Lib/sysconfig.py
@@ -527,6 +527,7 @@ def _init_posix(vars):
     name = _get_sysconfigdata_name()
     if 'STAGING_LIBDIR' in os.environ:
         sys.path.append(os.environ['STAGING_LIBDIR']+'/python-sysconfigdata')
+        sys.path.append(os.environ['STAGING_LIBDIR']+'/../lib64/python-sysconfigdata')
     _temp = __import__(name, globals(), locals(), ['build_time_vars'], 0)
     build_time_vars = _temp.build_time_vars
     vars.update(build_time_vars)