From f960d46dc4b626a4c3ceee3a196348cd548b494a Mon Sep 17 00:00:00 2001
From: BBgod118 <hewenbin11@h-partners.com>
Date: Tue, 10 Feb 2026 10:22:40 +0800
Subject: [PATCH] hi3093: irq-gic-v3 register SGI No.9 on virq No.10

---
 drivers/irqchip/irq-gic-v3.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/irqchip/irq-gic-v3.c b/drivers/irqchip/irq-gic-v3.c
index 0e85fcf91945..3c20bf4f6906 100644
--- a/drivers/irqchip/irq-gic-v3.c
+++ b/drivers/irqchip/irq-gic-v3.c
@@ -1310,7 +1310,7 @@ static void __init gic_smp_init(void)
 				  gic_starting_cpu, NULL);
 
 	/* Register all 8 non-secure SGIs */
-	base_sgi = __irq_domain_alloc_irqs(gic_data.domain, -1, 8,
+	base_sgi = __irq_domain_alloc_irqs(gic_data.domain, -1, 10,
 					   NUMA_NO_NODE, &sgi_fwspec,
 					   false, NULL);
 	if (WARN_ON(base_sgi <= 0))
-- 
2.34.1

