diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 8de1808..f209340 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -49,6 +49,7 @@ target_include_directories(libisulad_tools
 
 set_target_properties(libisulad_tools PROPERTIES PREFIX "")
 target_link_libraries(libisulad_tools ${ZLIB_LIBRARY} ${ISULA_LIBUTILS_LIBRARY} ${CRYPTO_LIBRARY})
+target_link_libraries(libisulad_tools -lgcompat)
 
 if (ENABLE_OCI_IMAGE)
     target_link_libraries(libisulad_tools ${LIBARCHIVE_LIBRARY})
diff --git a/src/cmd/isula/client_console.c b/src/cmd/isula/client_console.c
index 555f59b..1eb54e1 100644
--- a/src/cmd/isula/client_console.c
+++ b/src/cmd/isula/client_console.c
@@ -137,8 +137,7 @@ int create_console_fifos(bool attach_stdin, bool attach_stdout, bool attach_stde
         return -1;
     }
 
-    ret = snprintf(subpath, sizeof(subpath), "%s/%s-%u-%u", name, type, (unsigned int)getpid(),
-                   (unsigned int)pthread_self());
+    ret = snprintf(subpath, sizeof(subpath), "%s/%s-%u", name, type, (unsigned int)getpid());
     if (ret < 0 || (size_t)ret >= sizeof(subpath)) {
         ERROR("Path is too long");
         goto cleanup;
diff --git a/src/cmd/isulad/main.c b/src/cmd/isulad/main.c
index 0cdbfb5..1a87af7 100644
--- a/src/cmd/isulad/main.c
+++ b/src/cmd/isulad/main.c
@@ -40,6 +40,7 @@
 #include <string.h>
 #include <strings.h>
 #include <sys/time.h>
+#include <malloc_utils.h>
 #ifdef ENABLE_SUP_GROUPS
 #include <grp.h>
 #endif
diff --git a/src/daemon/executor/container_cb/execution.c b/src/daemon/executor/container_cb/execution.c
index c2a0bdb..7cc59aa 100644
--- a/src/daemon/executor/container_cb/execution.c
+++ b/src/daemon/executor/container_cb/execution.c
@@ -23,6 +23,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <libgen.h>
+#include <malloc_utils.h>
 
 #include <isula_libutils/container_config.h>
 #include <isula_libutils/container_config_v2.h>
@@ -474,9 +475,9 @@ static int container_start_prepare(container_t *cont, const container_start_requ
 static void handle_start_io_thread_by_cc(uint32_t cc, int sync_fd, pthread_t thread_id)
 {
     if (cc == ISULAD_SUCCESS) {
-        if (thread_id > 0) {
+        if (thread_id != NULL) {
             if (pthread_detach(thread_id) != 0) {
-                SYSERROR("Failed to detach 0x%lx", thread_id);
+                SYSERROR("Failed to detach");
             }
         }
         if (sync_fd >= 0) {
@@ -488,9 +489,9 @@ static void handle_start_io_thread_by_cc(uint32_t cc, int sync_fd, pthread_t thr
                 ERROR("Failed to write eventfd: %s", strerror(errno));
             }
         }
-        if (thread_id > 0) {
+        if (thread_id != NULL) {
             if (pthread_join(thread_id, NULL) != 0) {
-                ERROR("Failed to join thread: 0x%lx", thread_id);
+                ERROR("Failed to join thread");
             }
         }
         if (sync_fd >= 0) {
@@ -508,7 +509,7 @@ static int container_start_cb(const container_start_request *request, container_
     char *fifopath = NULL;
     container_t *cont = NULL;
     int sync_fd = -1;
-    pthread_t thread_id = 0;
+    pthread_t thread_id = NULL;
 
     DAEMON_CLEAR_ERRMSG();
 
diff --git a/src/daemon/executor/container_cb/execution_create.c b/src/daemon/executor/container_cb/execution_create.c
index e8f74f1..38ef072 100644
--- a/src/daemon/executor/container_cb/execution_create.c
+++ b/src/daemon/executor/container_cb/execution_create.c
@@ -24,6 +24,7 @@
 #include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
+#include <malloc_utils.h>
 
 #include <isula_libutils/log.h>
 #include <isula_libutils/auto_cleanup.h>
diff --git a/src/daemon/executor/container_cb/execution_information.c b/src/daemon/executor/container_cb/execution_information.c
index 2848022..caca444 100644
--- a/src/daemon/executor/container_cb/execution_information.c
+++ b/src/daemon/executor/container_cb/execution_information.c
@@ -39,6 +39,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <sys/utsname.h>
+#include <malloc_utils.h>
 
 #include "isula_libutils/log.h"
 #include "isulad_config.h"
diff --git a/src/daemon/executor/container_cb/execution_stream.c b/src/daemon/executor/container_cb/execution_stream.c
index 7e928cf..e29321b 100644
--- a/src/daemon/executor/container_cb/execution_stream.c
+++ b/src/daemon/executor/container_cb/execution_stream.c
@@ -63,7 +63,10 @@
 #include "utils_file.h"
 #include "utils_verify.h"
 
-#if defined (__ANDROID__) || defined(__MUSL__)
+#if defined (__ANDROID__)
+#undef SIG_CANCEL_SIGNAL
+#undef PTHREAD_CANCEL_ENABLE
+#undef PTHREAD_CANCEL_DISABLE
 #define SIG_CANCEL_SIGNAL     SIGUSR1
 #define PTHREAD_CANCEL_ENABLE    1
 #define PTHREAD_CANCEL_DISABLE    0
@@ -283,9 +286,9 @@ out:
 static void handle_attach_io_thread_by_cc(uint32_t cc, int sync_fd, pthread_t thread_id)
 {
     if (cc == ISULAD_SUCCESS) {
-        if (thread_id > 0) {
+        if (thread_id != NULL) {
             if (pthread_detach(thread_id) != 0) {
-                SYSERROR("Failed to detach 0x%lx", thread_id);
+                SYSERROR("Failed to detach");
             }
         }
         if (sync_fd >= 0) {
@@ -297,9 +300,9 @@ static void handle_attach_io_thread_by_cc(uint32_t cc, int sync_fd, pthread_t th
                 ERROR("Failed to write eventfd: %s", strerror(errno));
             }
         }
-        if (thread_id > 0) {
+        if (thread_id != NULL) {
             if (pthread_join(thread_id, NULL) != 0) {
-                ERROR("Failed to join thread: 0x%lx", thread_id);
+                ERROR("Failed to join thread");
             }
         }
         if (sync_fd >= 0) {
@@ -317,7 +320,7 @@ static int container_attach_cb(const container_attach_request *request, containe
     char *fifos[3] = { NULL, NULL, NULL };
     char *fifopath = NULL;
     int syncfd = -1;
-    pthread_t tid = 0;
+    pthread_t tid = NULL;
     container_t *cont = NULL;
     rt_attach_params_t params = { 0 };
 
diff --git a/src/daemon/executor/image_cb/image_cb.c b/src/daemon/executor/image_cb/image_cb.c
index 124feb2..ae00721 100644
--- a/src/daemon/executor/image_cb/image_cb.c
+++ b/src/daemon/executor/image_cb/image_cb.c
@@ -47,6 +47,7 @@
 #include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
+#include <malloc_utils.h>
 
 #include "utils.h"
 #include "error.h"
@@ -1215,4 +1216,4 @@ void image_callback_init(service_image_callback_t *cb)
 #ifdef ENABLE_IMAGE_SEARCH
     cb->search = image_search_cb;
 #endif
-}
\ No newline at end of file
+}
diff --git a/src/daemon/modules/events/monitord.c b/src/daemon/modules/events/monitord.c
index 775fb9a..b9d6eee 100644
--- a/src/daemon/modules/events/monitord.c
+++ b/src/daemon/modules/events/monitord.c
@@ -24,6 +24,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <sys/prctl.h>
+#include <malloc_utils.h>
 
 #include "isula_libutils/log.h"
 #include "mainloop.h"
diff --git a/src/daemon/modules/image/image_rootfs_handler.c b/src/daemon/modules/image/image_rootfs_handler.c
index 1275658..e6a6392 100644
--- a/src/daemon/modules/image/image_rootfs_handler.c
+++ b/src/daemon/modules/image/image_rootfs_handler.c
@@ -26,6 +26,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <pwd_utils.h>
 
 #include "isula_libutils/log.h"
 #include "err_msg.h"
diff --git a/src/daemon/modules/image/oci/storage/layer_store/layer_store.c b/src/daemon/modules/image/oci/storage/layer_store/layer_store.c
index 4edd0ca..a284145 100644
--- a/src/daemon/modules/image/oci/storage/layer_store/layer_store.c
+++ b/src/daemon/modules/image/oci/storage/layer_store/layer_store.c
@@ -32,6 +32,7 @@
 #include <sys/stat.h>
 #include <archive.h>
 #include <archive_entry.h>
+#include <unistd.h>
 
 #include "storage.h"
 #include "layer.h"
diff --git a/src/daemon/modules/service/io_handler.c b/src/daemon/modules/service/io_handler.c
index 98c763a..e02279c 100644
--- a/src/daemon/modules/service/io_handler.c
+++ b/src/daemon/modules/service/io_handler.c
@@ -116,7 +116,6 @@ int create_daemon_fifos(const char *id, const char *runtime, bool attach_stdin,
     char subpath[PATH_MAX] = { 0 };
     char fifodir[PATH_MAX] = { 0 };
     struct timespec now;
-    pthread_t tid;
 
     nret = clock_gettime(CLOCK_REALTIME, &now);
     if (nret != 0) {
@@ -124,15 +123,13 @@ int create_daemon_fifos(const char *id, const char *runtime, bool attach_stdin,
         goto cleanup;
     }
 
-    tid = pthread_self();
-
     statepath = conf_get_routine_statedir(runtime);
     if (statepath == NULL) {
         ERROR("State path is NULL");
         goto cleanup;
     }
 
-    nret = snprintf(subpath, PATH_MAX, "%s/%s/%u_%u_%u", id, operation, (unsigned int)tid, (unsigned int)now.tv_sec,
+    nret = snprintf(subpath, PATH_MAX, "%s/%s/%u_%u", id, operation,  (unsigned int)now.tv_sec,
                     (unsigned int)(now.tv_nsec));
     if (nret >= PATH_MAX || nret < 0) {
         ERROR("Failed to print string");
diff --git a/src/daemon/modules/service/service_container.c b/src/daemon/modules/service/service_container.c
index d69ee75..1110269 100644
--- a/src/daemon/modules/service/service_container.c
+++ b/src/daemon/modules/service/service_container.c
@@ -2064,9 +2064,9 @@ static void exec_container_end(container_exec_response *response, uint32_t cc, i
             ERROR("Failed to write eventfd: %s", strerror(errno));
         }
     }
-    if (thread_id > 0) {
+    if (thread_id != NULL) {
         if (pthread_join(thread_id, NULL) != 0) {
-            ERROR("Failed to join thread: 0x%lx", thread_id);
+            ERROR("Failed to join thread");
         }
     }
     if (sync_fd >= 0) {
@@ -2137,7 +2137,7 @@ int exec_container(const container_t *cont, const container_exec_request *reques
     char *id = NULL;
     char *fifos[3] = { NULL, NULL, NULL };
     char *fifopath = NULL;
-    pthread_t thread_id = 0;
+    pthread_t thread_id = NULL;
     defs_process_user *puser = NULL;
     char exec_command[EVENT_ARGS_MAX] = { 0x00 };
 
