diff --git a/etc/pam.d/chage b/etc/pam.d/chage
index 8f49f5c..b1f365d 100644
--- a/etc/pam.d/chage
+++ b/etc/pam.d/chage
@@ -1,4 +1,4 @@
 #%PAM-1.0
 auth		sufficient	pam_rootok.so
 account		required	pam_permit.so
-password	include		system-auth
+password	include		common-password
diff --git a/etc/pam.d/chgpasswd b/etc/pam.d/chgpasswd
index 8f49f5c..b1f365d 100644
--- a/etc/pam.d/chgpasswd
+++ b/etc/pam.d/chgpasswd
@@ -1,4 +1,4 @@
 #%PAM-1.0
 auth		sufficient	pam_rootok.so
 account		required	pam_permit.so
-password	include		system-auth
+password	include		common-password
diff --git a/etc/pam.d/groupadd b/etc/pam.d/groupadd
index 8f49f5c..b1f365d 100644
--- a/etc/pam.d/groupadd
+++ b/etc/pam.d/groupadd
@@ -1,4 +1,4 @@
 #%PAM-1.0
 auth		sufficient	pam_rootok.so
 account		required	pam_permit.so
-password	include		system-auth
+password	include		common-password
diff --git a/etc/pam.d/groupdel b/etc/pam.d/groupdel
index 8f49f5c..b1f365d 100644
--- a/etc/pam.d/groupdel
+++ b/etc/pam.d/groupdel
@@ -1,4 +1,4 @@
 #%PAM-1.0
 auth		sufficient	pam_rootok.so
 account		required	pam_permit.so
-password	include		system-auth
+password	include		common-password
diff --git a/etc/pam.d/groupmems b/etc/pam.d/groupmems
index 8f49f5c..b1f365d 100644
--- a/etc/pam.d/groupmems
+++ b/etc/pam.d/groupmems
@@ -1,4 +1,4 @@
 #%PAM-1.0
 auth		sufficient	pam_rootok.so
 account		required	pam_permit.so
-password	include		system-auth
+password	include		common-password
diff --git a/etc/pam.d/groupmod b/etc/pam.d/groupmod
index 8f49f5c..b1f365d 100644
--- a/etc/pam.d/groupmod
+++ b/etc/pam.d/groupmod
@@ -1,4 +1,4 @@
 #%PAM-1.0
 auth		sufficient	pam_rootok.so
 account		required	pam_permit.so
-password	include		system-auth
+password	include		common-password
diff --git a/etc/pam.d/useradd b/etc/pam.d/useradd
index 8f49f5c..b1f365d 100644
--- a/etc/pam.d/useradd
+++ b/etc/pam.d/useradd
@@ -1,4 +1,4 @@
 #%PAM-1.0
 auth		sufficient	pam_rootok.so
 account		required	pam_permit.so
-password	include		system-auth
+password	include		common-password
diff --git a/etc/pam.d/userdel b/etc/pam.d/userdel
index 8f49f5c..b1f365d 100644
--- a/etc/pam.d/userdel
+++ b/etc/pam.d/userdel
@@ -1,4 +1,4 @@
 #%PAM-1.0
 auth		sufficient	pam_rootok.so
 account		required	pam_permit.so
-password	include		system-auth
+password	include		common-password
diff --git a/etc/pam.d/usermod b/etc/pam.d/usermod
index 8f49f5c..b1f365d 100644
--- a/etc/pam.d/usermod
+++ b/etc/pam.d/usermod
@@ -1,4 +1,4 @@
 #%PAM-1.0
 auth		sufficient	pam_rootok.so
 account		required	pam_permit.so
-password	include		system-auth
+password	include		common-password
