diff -urpN a/compat/nss_compat.h b/compat/nss_compat.h
--- a/compat/nss_compat.h	2024-04-01 15:26:01.617300203 +0800
+++ b/compat/nss_compat.h	2024-04-01 15:24:17.730238759 +0800
@@ -23,9 +23,9 @@
 #ifndef COMPAT__NSS_H
 #define COMPAT__NSS_H 1
 
-#ifdef HAVE_NSS_H
-#include <nss.h>
-#endif /* HAVE_NSS_H */
+//#ifdef HAVE_NSS_H
+//#include <nss.h>
+//#endif /* HAVE_NSS_H */
 #ifdef HAVE_NSS_COMMON_H
 #include <nss_common.h>
 #endif /* HAVE_NSS_COMMON_H */
@@ -52,8 +52,19 @@
 
 #include "compat/ether.h"
 
+/* Possible results of lookup using a nss_* function.  */
+enum nss_status
+{
+  NSS_STATUS_TRYAGAIN = -2,
+  NSS_STATUS_UNAVAIL,
+  NSS_STATUS_NOTFOUND,
+  NSS_STATUS_SUCCESS,
+  NSS_STATUS_RETURN
+};
+
+
 /* define missing status codes */
-#ifndef HAVE_ENUM_NSS_STATUS
+/*#ifndef HAVE_ENUM_NSS_STATUS
 #ifndef NSS_STATUS_SUCCESS
 #define NSS_STATUS_SUCCESS NSS_SUCCESS
 #endif
@@ -69,12 +80,12 @@
 #ifndef NSS_STATUS_RETURN
 #define NSS_STATUS_RETURN NSS_NOTFOUND
 #endif
-#endif /* not HAVE_ENUM_NSS_STATUS */
+#endif*/ /* not HAVE_ENUM_NSS_STATUS */
 
 /* define nss_status_t */
-#ifdef HAVE_ENUM_NSS_STATUS
+//#ifdef HAVE_ENUM_NSS_STATUS
 typedef enum nss_status nss_status_t;
-#endif
+//#endif
 
 /* Define an aliasent if it was not found on the system. */
 #ifndef HAVE_STRUCT_ALIASENT
