--- originbot_navigation/launch/occupancy_grid.launch.py_org	2023-05-30 20:29:06.269595408 +0800
+++ originbot_navigation/launch/occupancy_grid.launch.py	2023-05-30 20:29:45.577828047 +0800
@@ -43,9 +43,9 @@
 
         Node(
             package='cartographer_ros',
-            executable='occupancy_grid_node',
+            executable='cartographer_occupancy_grid_node',
             name='occupancy_grid_node',
             output='screen',
             parameters=[{'use_sim_time': use_sim_time}],
             arguments=['-resolution', resolution, '-publish_period_sec', publish_period_sec]),
-    ])
\ No newline at end of file
+    ])
