diff -Naur originbot_base_org/src/originbot_base.cpp originbot_base/src/originbot_base.cpp
--- originbot_base_org/src/originbot_base.cpp	2023-05-31 12:01:13.355862106 +0800
+++ originbot_base/src/originbot_base.cpp	2023-05-31 12:00:50.429723711 +0800
@@ -20,17 +20,17 @@
 {
     // 加载参数
     std::string port_name="ttyS3";    
-    this->declare_parameter("port_name");           //声明及获取串口号参数
+    this->declare_parameter<std::string>("port_name", port_name);           //声明及获取串口号参数
     this->get_parameter_or<std::string>("port_name", port_name, "ttyS3");
-    this->declare_parameter("correct_factor_vx");   //声明及获取线速度校正参数
+    this->declare_parameter<float>("correct_factor_vx", 1.0);   //声明及获取线速度校正参数
     this->get_parameter_or<float>("correct_factor_vx", correct_factor_vx_, 1.0);
-    this->declare_parameter("correct_factor_vth");  //声明及获取角速度校正参数
+    this->declare_parameter<float>("correct_factor_vth", 1.0);  //声明及获取角速度校正参数
     this->get_parameter_or<float>("correct_factor_vth", correct_factor_vth_, 1.0);
-    this->declare_parameter("auto_stop_on");        //声明及获取自动停车功能的开关值
+    this->declare_parameter<bool>("auto_stop_on", true);        //声明及获取自动停车功能的开关值
     this->get_parameter_or<bool>("auto_stop_on", auto_stop_on_, true);
-    this->declare_parameter("use_imu");             //声明是否使用imu
+    this->declare_parameter<bool>("use_imu", false);             //声明是否使用imu
     this->get_parameter_or<bool>("use_imu", use_imu_, false);
-    this->declare_parameter("pub_odom");             //声明是否发布odom的tf
+    this->declare_parameter<bool>("pub_odom", false);             //声明是否发布odom的tf
     this->get_parameter_or<bool>("pub_odom", pub_odom_, false);
     
     // 打印加载的参数值
