diff --git a/setuptools/dist.py b/setuptools/dist.py
index 4458a58..3312b7b 100644
--- a/setuptools/dist.py
+++ b/setuptools/dist.py
@@ -216,7 +216,7 @@ def write_pkg_file(self, file):  # noqa: C901  # is too complex (14)  # FIXME
     if self.long_description_content_type:
         write_field('Description-Content-Type', self.long_description_content_type)
     if self.provides_extras:
-        for extra in self.provides_extras:
+        for extra in sorted(self.provides_extras):
             write_field('Provides-Extra', extra)
 
     self._write_list(file, 'License-File', self.license_files or [])
