--- rng-tools-6.14/rngd_jitter.c	2022-11-16 05:51:23.387022953 +0000
+++ rng-tools-6.14/rngd_jitter.c	2022-11-16 05:51:41.602597386 +0000
@@ -67,7 +67,7 @@ static int rngd_notime_start(void *ctx,
 	for(i=i-1;i>=0;i--) {
 		CPU_SET(i,cpus);
 	}
-	pthread_attr_setaffinity_np(&thread_ctx->notime_pthread_attr, cpusize, cpus);
+	pthread_setaffinity_np(&thread_ctx->notime_pthread_attr, cpusize, cpus);
 
 	ret = -pthread_create(&thread_ctx->notime_thread_id,
 				&thread_ctx->notime_pthread_attr,
