diff --git a/libmetal/cmake/syscheck.cmake b/libmetal/cmake/syscheck.cmake
index 032a849..303a0e8 100644
--- a/libmetal/cmake/syscheck.cmake
+++ b/libmetal/cmake/syscheck.cmake
@@ -1,7 +1,7 @@
 # use "Generic" as CMAKE_SYSTEM_NAME
 
 if (WITH_ZEPHYR)
-  set (CMAKE_SYSTEM_NAME "Generic"             CACHE STRING "")
+  set (CMAKE_SYSTEM_NAME "Generic")
   string (TOLOWER "Zephyr"                PROJECT_SYSTEM)
   string (TOUPPER "Zephyr"                PROJECT_SYSTEM_UPPER)
   if (NOT WITH_ZEPHYR_LIB)
