From cdf2d060b3d66ed1a8c86ee22a3d019a8ce6056e Mon Sep 15 00:00:00 2025
From: oee <oee@oee.com>
Date: Mon, 15 Sep 2025 18:59:45 +0800
Subject: [PATCH] yocto: fix 6.6 kernel outree module + ieeuer. 

fix 6.6 kernel outree module + ieeuer in yocto build.

Signed-off-by: oee <oee@oee.com>
--- a/scripts/setlocalversion
+++ b/scripts/setlocalversion
@@ -113,7 +113,6 @@ scm_version()
 		# If only the short version is requested, don't bother
 		# running further git commands
 		if $short; then
-			echo "+"
 			return
 		fi
 
