diff --git a/sample/encoder_decoder_demo/CMakeLists.txt b/sample/encoder_decoder_demo/CMakeLists.txt
index 956a46e..2d3c74f 100644
--- a/sample/encoder_decoder_demo/CMakeLists.txt
+++ b/sample/encoder_decoder_demo/CMakeLists.txt
@@ -4,7 +4,7 @@ cmake_minimum_required(VERSION 3.0)
 # 项目名称
 project(encoder_decoder_demo)
 
-set(CMAKE_C_COMPILER aarch64-mix210-linux-gcc)
+#set(CMAKE_C_COMPILER aarch64-mix210-linux-gcc)
 
 # 添加可执行文件
 add_executable(encoder_decoder_demo hibot_bot_test.c)
@@ -19,14 +19,14 @@ target_include_directories(encoder_decoder_demo
 
 target_link_libraries(encoder_decoder_demo
 PUBLIC
-  $ENV{SDK_PATH}/smp/a55_linux/mpp/out/lib/libsecurec.so
-  $ENV{SDK_PATH}/smp/a55_linux/mpp/out/lib/libss_dsp.so
-  $ENV{SDK_PATH}/smp/a55_linux/mpp/out/lib/libss_mpi.so
-  $ENV{SDK_PATH}/smp/a55_linux/mpp/out/lib/libss_dnvqe.so
-  $ENV{SDK_PATH}/smp/a55_linux/mpp/out/lib/libss_upvqe.so
-  $ENV{SDK_PATH}/smp/a55_linux/mpp/out/lib/libss_voice_engine.so
-  $ENV{SDK_PATH}/smp/a55_linux/mpp/out/lib/libss_hdmi.so
-  ${CMAKE_CURRENT_SOURCE_DIR}/lib/libhibot.so
+  securec
+  ss_dsp
+  ss_mpi
+  ss_dnvqe
+  ss_upvqe
+  ss_voice_engine
+  ss_hdmi
+  hibot
   m
   pthread
 )
