--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -32,8 +32,16 @@ ament_target_dependencies(hw_calc_demo
   image_transport
   image_geometry
 )
-target_link_libraries(hw_calc_demo
-  ${CMAKE_CURRENT_SOURCE_DIR}/3rd/lib/libhibot.so
+
+target_link_libraries(hw_calc_demo 
+  hibot
+  securec
+  ss_dsp
+  ss_mpi
+  ss_dnvqe
+  ss_upvqe
+  ss_voice_engine
+  ss_hdmi
 )
 
 # Install targets
