--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -19,7 +19,16 @@ link_directories(
 )
 
 add_executable(test_recv src/test_recv.cpp)
-target_link_libraries(test_recv hibot)
+target_link_libraries(test_recv 
+  securec
+  ss_dsp
+  ss_mpi
+  ss_dnvqe
+  ss_upvqe
+  ss_voice_engine
+  ss_hdmi
+  hibot
+)
 ament_target_dependencies(test_recv rclcpp shm_meta_info)
 
 install(
@@ -34,4 +43,4 @@ install(
   FILES ../SHMBufferRosPkg/lib/libhibot.so
   DESTINATION lib)
 
-ament_package()
\ No newline at end of file
+ament_package()
