--- qt-apps-qtwebbrowser-5.15.1_QtAS/src/src.pro	2020-05-06 16:48:56.000000000 +0800
+++ a/src/src.pro	2024-04-24 16:42:53.285883453 +0800
@@ -40,7 +40,7 @@ RESOURCES += resources.qrc
     isEmpty(INSTALL_PREFIX): INSTALL_PREFIX=/usr/local/bin
 } else {
     # Path for Qt for Device Creation
-    isEmpty(INSTALL_PREFIX): INSTALL_PREFIX=/data/user/qt/qtwebbrowser-app
+    isEmpty(INSTALL_PREFIX): INSTALL_PREFIX=$$(bindir)
 }
 
 target.path = $$INSTALL_PREFIX
