diff -uprN a/toolchain/linux/BUILD.gn  b/toolchain/linux/BUILD.gn
--- a/toolchain/linux/BUILD.gn  2022-09-14 16:44:09.577085155 +0800
+++ b/toolchain/linux/BUILD.gn  2022-09-14 16:47:56.430068282 +0800
@@ -23,7 +23,7 @@ clang_toolchain("clang_arm64") {

 gcc_toolchain("arm64") {
   tool_path = rebase_path("//toolchain/bin", root_build_dir)
-  toolprefix = "${tool_path}/aarch64-openeuler-linux-gnu-"
+  toolprefix = "${tool_path}/aarch64-openeuler-linux-musl-"

   cc = "${toolprefix}gcc"
   cxx = "${toolprefix}g++"
diff --git a/config/compiler/BUILD.gn b/config/compiler/BUILD.gn
index 14c07bb..d5766e1 100755
--- a/config/compiler/BUILD.gn
+++ b/config/compiler/BUILD.gn
@@ -864,6 +864,7 @@ config("default_warnings") {
     # Don't warn about "maybe" uninitialized. Clang doesn't include this
     # in -Wall but gcc does, and it gives false positives.
     cflags += [ "-Wno-maybe-uninitialized" ]
+    cflags += [ "-Wno-error" ]
     cflags += [ "-Wno-deprecated-declarations" ]

     # GCC assumes 'this' is never nullptr and optimizes away code
