diff -uprN a/toolchain/linux/BUILD.gn  b/toolchain/linux/BUILD.gn
--- a/toolchain/linux/BUILD.gn  2022-09-14 16:44:09.577085155 +0800
+++ b/toolchain/linux/BUILD.gn  2022-09-14 16:47:56.430068282 +0800
@@ -23,7 +23,7 @@ clang_toolchain("clang_arm64") {

 gcc_toolchain("arm64") {
   tool_path = rebase_path("//toolchain/bin", root_build_dir)
-  toolprefix = "${tool_path}/aarch64-openeuler-linux-gnu-"
+  toolprefix = "${tool_path}/aarch64-openeuler-linux-musl-"

   cc = "${toolprefix}gcc"
   cxx = "${toolprefix}g++"
