diff --git a/config/compiler/BUILD.gn b/config/compiler/BUILD.gn
index d5766e1..03bb10d 100755
--- a/config/compiler/BUILD.gn
+++ b/config/compiler/BUILD.gn
@@ -863,7 +863,7 @@ config("default_warnings") {
 
     # Don't warn about "maybe" uninitialized. Clang doesn't include this
     # in -Wall but gcc does, and it gives false positives.
-    cflags += [ "-Wno-maybe-uninitialized" ]
+    #cflags += [ "-Wno-maybe-uninitialized" ]
     cflags += [ "-Wno-error" ]
     cflags += [ "-Wno-deprecated-declarations" ]
 
diff --git a/toolchain/linux/BUILD.gn b/toolchain/linux/BUILD.gn
index d852c12..2e7af39 100755
--- a/toolchain/linux/BUILD.gn
+++ b/toolchain/linux/BUILD.gn
@@ -22,11 +22,10 @@ clang_toolchain("clang_arm64") {
 }
 
 gcc_toolchain("arm64") {
-  tool_path = rebase_path("//toolchain/bin", root_build_dir)
-  toolprefix = "${tool_path}/aarch64-openeuler-linux-musl-"
+  toolprefix = "aarch64-openeuler-linux-musl-"
 
-  cc = "${toolprefix}gcc"
-  cxx = "${toolprefix}g++"
+  cc = "${toolprefix}clang"
+  cxx = "${toolprefix}clang++"
 
   ar = "${toolprefix}ar"
   ld = cxx
