diff --git a/cmake/set_build_flags.cmake b/cmake/set_build_flags.cmake
index 3806979..d6acdd4 100755
--- a/cmake/set_build_flags.cmake
+++ b/cmake/set_build_flags.cmake
@@ -23,6 +23,11 @@ if (NOT DISABLE_WERROR)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror")
 endif()
 
+if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-error=return-type-c-linkage -Wno-error=unqualified-std-cast-call -Wno-error=unused-private-field -Wno-error=missing-braces -Wno-error=non-pod-varargs -Wno-error=pessimizing-move -Wno-error=format-security -Wno-error=format-insufficient-args")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-error=return-type-c-linkage -Wno-error=unqualified-std-cast-call -Wno-error=unused-private-field -Wno-error=missing-braces -Wno-error=non-pod-varargs -Wno-error=pessimizing-move -Wno-error=format-security -Wno-error=format-insufficient-args")
+endif()
+
 if (ISULAD_GCOV)
     set(CMAKE_C_FLAGS_DEBUG "-Wall -fprofile-arcs -ftest-coverage")
     set(CMAKE_CXX_FLAGS_DEBUG "-Wall -fprofile-arcs -ftest-coverage -Wno-error=deprecated-declarations")
--

