From da43b7662d2653f30e3d91c78558d738ab2ae488 Mon Sep 17 00:00:00 2001
From: s_c_c <shichuchao@huawei.com>
Date: Mon, 25 Mar 2024 17:08:11 +0800
Subject: [PATCH] feat for embedded eventhandler comment out cflags

---
 frameworks/native/BUILD.gn | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/frameworks/native/BUILD.gn b/frameworks/native/BUILD.gn
index 6c98a96..9dd78dc 100644
--- a/frameworks/native/BUILD.gn
+++ b/frameworks/native/BUILD.gn
@@ -24,7 +24,7 @@ ohos_shared_library("eventhandler_native") {
     "${sysroot}/usr/include/hilog",
   ]
 
-  cflags = [ "-Wno-error=inconsistent-missing-override" ]
+#   cflags = [ "-Wno-error=inconsistent-missing-override" ]
 
   deps = [ "${frameworks_path}/eventhandler:libeventhandler" ]
 
-- 
2.20.1 (Apple Git-117)

