--- a/bl31/aarch64/bl31_entrypoint.S	2024-07-31 08:43:14.814966080 +0000
+++ b/bl31/aarch64/bl31_entrypoint.S	2024-07-31 08:56:12.780141461 +0000
@@ -27,8 +27,8 @@ func bl31_entrypoint
 	 * Stash the previous bootloader arguments x0 - x3 for later use.
 	 * ---------------------------------------------------------------
 	 */
-	ldr x0,  =OS_SYS_CTRL_REG2
-	ldr x1,  =OS_SYS_CTRL_REG4
+	ldr x0,  =0x11020308
+	ldr x1,  =0x11020310
 	ldr x20, [x0]
 	ldr x21, [x1]
 	mov x22, #0
