--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -41,19 +44,24 @@ find_package(tf2_ros REQUIRED)
 find_package(tf2_sensor_msgs REQUIRED)
 find_package(PkgConfig REQUIRED)
 
-pkg_search_module(LIBUVC REQUIRED libuvc)
-if (NOT LIBUVC_FOUND)
-  message(FATAL_ERROR "libuvc is not found")
-endif ()
-pkg_search_module(GLOG REQUIRED libglog)
+find_package(libuvc REQUIRED)
+find_package(glog REQUIRED)
 
-if (NOT GLOG_FOUND)
-  message(FATAL_ERROR "glog is not found")
-endif ()
+#pkg_search_module(LIBUVC REQUIRED libuvc)
+#if (NOT LIBUVC_FOUND)
+#  message(FATAL_ERROR "libuvc is not found")
+#endif ()
+#pkg_search_module(GLOG REQUIRED libglog)
+
+#if (NOT GLOG_FOUND)
+#  message(FATAL_ERROR "glog is not found")
+#endif ()
 
 
-execute_process(COMMAND uname -m OUTPUT_VARIABLE MACHINES)
-execute_process(COMMAND getconf LONG_BIT OUTPUT_VARIABLE MACHINES_BIT)
+#execute_process(COMMAND uname -m OUTPUT_VARIABLE MACHINES)
+#execute_process(COMMAND getconf LONG_BIT OUTPUT_VARIABLE MACHINES_BIT)
+set(MACHINES aarch64)
+set(MACHINES_BIT 64)
 message(STATUS "ORRBEC Machine : ${MACHINES}")
 message(STATUS "ORRBEC Machine Bits : ${MACHINES_BIT}")
 
@@ -129,6 +137,8 @@ target_link_libraries(${PROJECT_NAME}
   Eigen3::Eigen
   ${GLOG_LIBRARIES}
   ${LIBUVC_LIBRARIES}
+  -luvc
+  -lglog
   -lOpenNI2
   -L${ORBBEC_OPENNI2_REDIST}
   )
@@ -157,6 +167,8 @@ target_link_libraries(${PROJECT_NAME}_node
   ${GLOG_LIBRARIES}
   ${LIBUVC_LIBRARIES}
   -lOpenNI2
+  -luvc
+  -lglog
   -L${ORBBEC_OPENNI2_REDIST}
   ${PROJECT_NAME}
   )
@@ -181,6 +193,8 @@ target_link_libraries(list_devices_node
   Eigen3::Eigen
   ${GLOG_LIBRARIES}
   ${LIBUVC_LIBRARIES}
+  -luvc
+  -lglog
   -lOpenNI2
   -L${ORBBEC_OPENNI2_REDIST}
   ${PROJECT_NAME}
@@ -208,6 +222,8 @@ target_link_libraries(clean_shm_node
   Eigen3::Eigen
   ${GLOG_LIBRARIES}
   ${LIBUVC_LIBRARIES}
+  -luvc
+  -lglog
   -lOpenNI2
   -L${ORBBEC_OPENNI2_REDIST}
   ${PROJECT_NAME}
