diff --git a/bsp/qemu-virt64-aarch64/rtconfig.py b/bsp/qemu-virt64-aarch64/rtconfig.py
index 7204079e6..7f2ea47f6 100644
--- a/bsp/qemu-virt64-aarch64/rtconfig.py
+++ b/bsp/qemu-virt64-aarch64/rtconfig.py
@@ -27,7 +27,13 @@ BUILD = 'debug'
 
 if PLATFORM == 'gcc':
     # toolchains
-    PREFIX = 'aarch64-none-elf-'
+    if os.getenv('OPENEULER_RTT_GCC_PREFIX'):
+        PREFIX = os.getenv('OPENEULER_RTT_GCC_PREFIX')
+        TARGET_TUNE = os.getenv('OPENEULER_RTT_GCC_FLAGS') + ' -g -march=armv8-a -mtune=cortex-a57'
+    else:
+        PREFIX = 'aarch64-elf-'
+        TARGET_TUNE = ' -g -march=armv8-a -mtune=cortex-a53'
+
     CC      = PREFIX + 'gcc'
     CXX     = PREFIX + 'g++'
     AS      = PREFIX + 'gcc'
@@ -38,7 +44,7 @@ if PLATFORM == 'gcc':
     OBJDUMP = PREFIX + 'objdump'
     OBJCPY  = PREFIX + 'objcopy'
 
-    DEVICE = ' -g -march=armv8-a -mtune=cortex-a53'
+    DEVICE = TARGET_TUNE
     CFLAGS = DEVICE + ' -Wall'
     AFLAGS = ' -c' + ' -x assembler-with-cpp -D__ASSEMBLY__'
     LFLAGS  = DEVICE + ' -nostartfiles -Wl,--gc-sections,-Map=rtthread.map,-cref,-u,system_vectors -T link.lds'
