commit 05f1ae3dac69f1c677d3d8d0fabfb1536b8c2583
Author: lisimin <lisimin@huawei.com>
Date:   Wed Sep 22 09:34:22 2021 +0800

    support arm32
    
    Signed-off-by: lisimin <lisimin@huawei.com>

diff --git a/configure.ac b/configure.ac
index 9eb6dcb..929d670 100644
--- a/configure.ac
+++ b/configure.ac
@@ -812,6 +812,7 @@ AM_CONDITIONAL([HAVE_ISULAD], [test "x$adapt_isulad" = "xyes"])
 if test "x$adapt_isulad" = "xyes"; then
 	AC_DEFINE([HAVE_ISULAD], 1, [adapt to iSulad])
 	AC_MSG_RESULT([yes])
+	AC_SUBST([YAJL_LIBS], [-lyajl])
 else
 	AC_MSG_RESULT([no])
 fi
diff --git a/src/lxc/json/json_common.c b/src/lxc/json/json_common.c
index ec20c59..87895b0 100755
--- a/src/lxc/json/json_common.c
+++ b/src/lxc/json/json_common.c
@@ -764,11 +764,11 @@ int append:json_map_int_string(json_map_int_string *map, int key, const char *va
     if (map == NULL) {
         return -1;
     }
-
+#if 0
     if ((SIZE_MAX / sizeof(int) - 1) < map->len || (SIZE_MAX / sizeof(char *) - 1) < map->len) {
         return -1;
     }
-
+#endif
     len = map->len + 1;
     keys = safe_malloc(len * sizeof(int));
     vals = safe_malloc(len * sizeof(char *));
@@ -886,11 +886,11 @@ int append:json_map_string_int(json_map_string_int *map, const char *key, int va
     if (map == NULL) {
         return -1;
     }
-
+#if 0
     if ((SIZE_MAX / sizeof(char *) - 1) < map->len || (SIZE_MAX / sizeof(int) - 1) < map->len) {
         return -1;
     }
-
+#endif
     len = map->len + 1;
     keys = safe_malloc(len * sizeof(char *));
     vals = safe_malloc(len * sizeof(int));
