Suject: Fix glib2 compile error for arm32
Signed-off-by: Zheng Liming <zhengliming3@huawei.com>

diff --git a/glib/gthreadprivate.h b/glib/gthreadprivate.h
index 6eaf422..0a1bb97 100644
--- a/glib/gthreadprivate.h
+++ b/glib/gthreadprivate.h
@@ -44,6 +44,7 @@ struct  _GRealThread
 #include <linux/futex.h>
 #include <sys/syscall.h>
 #include <unistd.h>
+#include <errno.h>

 #ifndef FUTEX_WAIT_PRIVATE
 #define FUTEX_WAIT_PRIVATE FUTEX_WAIT
