--- a/master/globals.h	2022-11-16 06:47:21.199110038 +0000
+++ b/master/globals.h	2022-11-16 06:48:10.055056280 +0000
@@ -259,6 +259,8 @@ extern char *ec_master_version_str;
 
 /*****************************************************************************/
 
+typedef long ssize_t; //for ssize_t
+
 unsigned int ec_master_count(void);
 void ec_print_data(const uint8_t *, size_t);
 void ec_print_data_diff(const uint8_t *, const uint8_t *, size_t);
