--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -44,8 +47,10 @@ set(MPI_DIR_LIB
 set(MPI_DIR_INCLUDED_DIRS
   "${CMAKE_CURRENT_SOURCE_DIR}/3rd/mpi/include/")
 
-execute_process(COMMAND uname -m OUTPUT_VARIABLE MACHINES)
-execute_process(COMMAND getconf LONG_BIT OUTPUT_VARIABLE MACHINES_BIT)
+#execute_process(COMMAND uname -m OUTPUT_VARIABLE MACHINES)
+#execute_process(COMMAND getconf LONG_BIT OUTPUT_VARIABLE MACHINES_BIT)
+set(MACHINES aarch64)
+set(MACHINES_BIT 64)
 message(STATUS "ORRBEC Machine : ${MACHINES}")
 message(STATUS "ORRBEC Machine Bits : ${MACHINES_BIT}")
 
