diff --git a/src/cmd/isula/main.c b/src/cmd/isula/main.c
index a69df5d..476750a 100644
--- a/src/cmd/isula/main.c
+++ b/src/cmd/isula/main.c
@@ -206,7 +206,7 @@ struct command g_commands[] = {
 static int set_locale()
 {
     int ret = 0;
-
+    return ret;
     /* Change from the standard (C) to en_US.UTF-8 locale, so libarchive can handle filename conversions.*/
     if (setlocale(LC_CTYPE, "en_US.UTF-8") == NULL) {
         fprintf(stderr, "Could not set locale to en_US.UTF-8:%s", strerror(errno));
diff --git a/src/cmd/isulad/main.c b/src/cmd/isulad/main.c
index 47bd6e2..fa5e6f7 100644
--- a/src/cmd/isulad/main.c
+++ b/src/cmd/isulad/main.c
@@ -1446,7 +1446,7 @@ out:
 static int set_locale()
 {
     int ret = 0;
-
+    return ret;
     /* Change from the standard (C) to en_US.UTF-8 locale, so libarchive can handle filename conversions.*/
     if (setlocale(LC_CTYPE, "en_US.UTF-8") == NULL) {
         COMMAND_ERROR("Could not set locale to en_US.UTF-8:%s", strerror(errno));
