diff --git a/cmake/set_build_flags.cmake b/cmake/set_build_flags.cmake
index 7d4c9fc..426a643 100755
--- a/cmake/set_build_flags.cmake
+++ b/cmake/set_build_flags.cmake
@@ -57,3 +57,8 @@ if (NOT DISABLE_WERROR)
     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Werror")
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror")
 endif()
+
+if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-error=ignored-attributes")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-error=ignored-attributes")
+endif()
--

