diff -uprN ros-humble-orocos-kdl-3.3.3/src/CMakeLists.txt ros-humble-orocos-kdl-3.3.3-new/src/CMakeLists.txt
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -104,7 +104,7 @@ ENDIF()
 ADD_LIBRARY(orocos-kdl ${LIB_TYPE} ${KDL_SRCS} "${CMAKE_CURRENT_BINARY_DIR}/config.h")
 
 TARGET_INCLUDE_DIRECTORIES(orocos-kdl PUBLIC
-  "$<INSTALL_INTERFACE:include>")
+  "$<INSTALL_INTERFACE:include/${PROJECT_NAME}>")
 
 SET_TARGET_PROPERTIES( orocos-kdl PROPERTIES
   SOVERSION "${KDL_VERSION_MAJOR}.${KDL_VERSION_MINOR}"
@@ -138,10 +138,10 @@ INSTALL(TARGETS orocos-kdl
   EXPORT OrocosKDLTargets
   ARCHIVE DESTINATION lib${LIB_SUFFIX}
   LIBRARY DESTINATION lib${LIB_SUFFIX}
-  PUBLIC_HEADER DESTINATION include/kdl
+  PUBLIC_HEADER DESTINATION include/${PROJECT_NAME}/kdl
 )
 
-INSTALL(FILES ${UTIL_HPPS} DESTINATION include/kdl/utilities)
+INSTALL(FILES ${UTIL_HPPS} DESTINATION include/${PROJECT_NAME}/kdl/utilities)
 
 # Orocos convention:
 CONFIGURE_FILE( kdl.pc.in ${CMAKE_CURRENT_BINARY_DIR}/orocos-kdl.pc @ONLY)
