--- pcl-pcl-1.12.1/PCLConfig.cmake.in_org	2023-05-30 08:25:44.499647259 +0800
+++ pcl-pcl-1.12.1/PCLConfig.cmake.in	2023-05-30 08:26:03.906767882 +0800
@@ -390,7 +390,7 @@
   endif()
 else()
 # PCLConfig.cmake is installed to PCL_ROOT/share/pcl-x.y
-  set(PCL_ROOT "@CMAKE_INSTALL_PREFIX@")
+  get_filename_component(PCL_ROOT "${CMAKE_CURRENT_LIST_DIR}/../.." ABSOLUTE)
 endif()
 
 # check whether PCLConfig.cmake is found into a PCL installation or in a build tree
