diff -uprN a/meson.build b/meson.build
--- a/meson.build	2023-03-17 02:12:01.119365345 +0000
+++ b/meson.build	2023-03-17 02:10:34.756309801 +0000
@@ -448,7 +448,6 @@ decl_headers = '''
 #include <uchar.h>
 #include <sys/mount.h>
 #include <sys/stat.h>
-#include <linux/fs.h>
 '''
 
 foreach decl : ['char16_t',
@@ -460,6 +459,17 @@ foreach decl : ['char16_t',
         # We get -1 if the size cannot be determined
         have = cc.sizeof(decl, prefix : decl_headers, args : '-D_GNU_SOURCE') > 0
 
+        if decl == 'struct mount_attr'
+                if have
+                        want_linux_fs_h = false
+                else
+                        have = cc.sizeof(decl,
+                                         prefix : decl_headers + '#include <linux/fs.h>',
+                                         args : '-D_GNU_SOURCE') > 0
+                        want_linux_fs_h = have
+                endif
+        endif
+
         if decl == 'struct statx'
                 if have
                         want_linux_stat_h = false
@@ -475,6 +485,7 @@ foreach decl : ['char16_t',
 endforeach
 
 conf.set10('WANT_LINUX_STAT_H', want_linux_stat_h)
+conf.set10('WANT_LINUX_FS_H', want_linux_fs_h)
 
 foreach ident : ['secure_getenv', '__secure_getenv']
         conf.set10('HAVE_' + ident.to_upper(), cc.has_function(ident))
diff -uprN a/src/core/namespace.c b/src/core/namespace.c
--- a/src/core/namespace.c	2023-03-17 02:11:54.819431730 +0000
+++ b/src/core/namespace.c	2023-03-17 02:10:34.772309619 +0000
@@ -6,7 +6,9 @@
 #include <stdio.h>
 #include <sys/mount.h>
 #include <unistd.h>
+#if WANT_LINUX_FS_H
 #include <linux/fs.h>
+#endif
 
 #include "alloc-util.h"
 #include "base-filesystem.h"
diff -uprN a/src/shared/mount-util.c b/src/shared/mount-util.c
--- a/src/shared/mount-util.c	2023-03-17 02:11:59.291384569 +0000
+++ b/src/shared/mount-util.c	2023-03-17 02:10:34.808309208 +0000
@@ -7,7 +7,9 @@
 #include <sys/statvfs.h>
 #include <unistd.h>
 #include <linux/loop.h>
+#if WANT_LINUX_FS_H
 #include <linux/fs.h>
+#endif
 
 #include "alloc-util.h"
 #include "dissect-image.h"
