diff -uprN a/src/basic/dirent-util.c b/src/basic/dirent-util.c
--- a/src/basic/dirent-util.c	2024-03-29 15:51:41.290624365 +0800
+++ b/src/basic/dirent-util.c	2024-03-29 16:08:48.748961186 +0800
@@ -82,7 +82,7 @@ bool dirent_is_journal_subdir(const stru
         if (!e)
                 return id128_is_valid(de->d_name); /* No namespace */
 
-        n = strndupa(de->d_name, e - de->d_name);
+        n = strndup(de->d_name, e - de->d_name);
         if (!id128_is_valid(n))
                 return false;
 
Binary files a/tools/chromiumos/__pycache__/gen_autosuspend_rules.cpython-310.pyc and b/tools/chromiumos/__pycache__/gen_autosuspend_rules.cpython-310.pyc differ
