diff --git a/alloc.c b/alloc.c
index cf7eb40..beda681 100644
--- a/alloc.c
+++ b/alloc.c
@@ -245,6 +245,7 @@ void free_huge_pages(void *ptr)
 	__free_huge_pages(ptr, 1);
 }
 
+#define _SC_LEVEL2_CACHE_LINESIZE 1150
 static size_t get_cacheline_size() {
 #if defined(__linux__)
 	FILE * fp = fopen("/sys/devices/system/cpu/cpu0/cache/index0/coherency_line_size", "r");
diff --git a/shm.c b/shm.c
index 9447b63..27b6ea7 100644
--- a/shm.c
+++ b/shm.c
@@ -29,6 +29,7 @@
 #include "libhugetlbfs_internal.h"
 #include "hugetlbfs.h"
 #include <sys/syscall.h>
+#include <sys/cdefs.h>
 
 #if defined(SYS_shmget) || defined(SYS_ipc)
 #define HAVE_SHMGET_SYSCALL
diff --git a/tests/hugetests.h b/tests/hugetests.h
index fbe4dc0..c3fa320 100644
--- a/tests/hugetests.h
+++ b/tests/hugetests.h
@@ -24,6 +24,7 @@
 #include <string.h>
 #include <dirent.h>
 #include <unistd.h>
+#include <sys/types.h>
 
 #include "libhugetlbfs_privutils.h"
 #include "libhugetlbfs_testprobes.h"
