--- LVM2.2.03.14/tools/Makefile.in.org	2022-12-05 06:02:57.551639891 +0000
+++ LVM2.2.03.14/tools/Makefile.in	2022-12-05 06:00:59.752750543 +0000
@@ -123,7 +123,7 @@ CFLAGS_lvm.o += $(EXTRA_EXEC_CFLAGS)
 lvm: $(OBJECTS) lvm.o $(LVMINTERNAL_LIBS)
 	@echo "    [CC] $@"
 	$(Q) $(CC) $(CFLAGS) $(LDFLAGS) $(EXTRA_EXEC_LDFLAGS) $(ELDFLAGS) \
-		-o $@ $+ $(LVMLIBS)
+		-o $@ $+ $(LVMLIBS) -lgcompat
 
 DEFS_man-generator.o += -DMAN_PAGE_GENERATOR
 
