--- a/src/3rdparty/gn/tools/gn/setup.cc
+++ b/src/3rdparty/gn/tools/gn/setup.cc
@@ -647,7 +647,8 @@ bool Setup::FillBuildDir(const std::string& build_dir, bool require_exists) {
         "I could not get the real path of \"" + FilePathToUTF8(build_dir_path) +
             "\".")
         .PrintToStdout();
-    return false;
+    build_dir_realpath = build_dir_path;
+    //return false;
   }
   resolved = SourceDirForPath(build_settings_.root_path(), build_dir_realpath);
 
