diff --git a/config/init/systemd/lxcfs.service.in b/config/init/systemd/lxcfs.service.in
index c3bfd0f..167b20d 100644
--- a/config/init/systemd/lxcfs.service.in
+++ b/config/init/systemd/lxcfs.service.in
@@ -7,6 +7,7 @@ Documentation=man:lxcfs(1)
 [Service]
 ExecStartPre=-/usr/local/bin/lxcfs-tools prestart
+ExecStartPre=-/bin/mkdir /var/lib/lxc/lxcfs
 ExecStart=/usr/bin/lxcfs --enable-cfs /var/lib/lxc/lxcfs/
 ExecStartPost=-/usr/local/bin/lxcfs-tools remount -a
 KillMode=process
 ExecStop=-/usr/local/bin/lxcfs-tools umount -a
