diff -uprN git/src/main.cpp git_new/src/main.cpp
--- git/src/main.cpp	2023-10-16 07:48:33.086569522 +0000
+++ git_new/src/main.cpp	2023-10-16 09:19:15.671121395 +0000
@@ -58,10 +58,10 @@ static bool drop_permissions(void)
 
     if ((setgid(0) != -1) || (setuid(0) != -1))
     {
-        LOGE("Unable to drop root (we shouldn't be able to "
+        LOGD("Unable to drop root (we shouldn't be able to "
              "restore it after setuid), refusing to start");
 
-        return false;
+        //Temporary permission to run with root
     }
 
     return true;
