diff --git a/lib/mm/memlock.c b/lib/mm/memlock.c
index ad69f64..8117719 100644
--- a/lib/mm/memlock.c
+++ b/lib/mm/memlock.c
@@ -27,6 +27,9 @@
 #include <sys/resource.h>
 #include <malloc.h>
 
+/*add header file for musl*/
+#include <malloc_utils.h>
+
 #ifdef HAVE_VALGRIND
 #include <valgrind.h>
 #endif
diff --git a/tools/vgimportdevices.c b/tools/vgimportdevices.c
index ab359d3..cc85068 100644
--- a/tools/vgimportdevices.c
+++ b/tools/vgimportdevices.c
@@ -16,6 +16,9 @@
 #include "lib/cache/lvmcache.h"
 #include "lib/device/device_id.h"
 
+/* coverity[unnecessary_header] needed for MuslC */
+#include <sys/file.h>
+
 struct vgimportdevices_params {
 	uint32_t added_devices;
 };
