From 3dbbfba1181b69588a544f782e3102d5626a7422 Mon Sep 17 00:00:00 2001
From: Yunfei Li <liyunfei33@huawei.com>
Date: Fri, 4 Nov 2022 06:50:50 +0000
Subject: [PATCH] remove asprintf declaration for clang build

---
 client/dhclient.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/client/dhclient.c b/client/dhclient.c
index 3b79bd1..cedb19f 100644
--- a/client/dhclient.c
+++ b/client/dhclient.c
@@ -49,7 +49,9 @@
  * Defined in stdio.h when _GNU_SOURCE is set, but we don't want to define
  * that when building ISC code.
  */
+#ifndef HAVE_ASPRINTF
 extern int asprintf(char **strp, const char *fmt, ...);
+#endif
 
 TIME default_lease_time = 43200; /* 12 hours... */
 TIME max_lease_time = 86400; /* 24 hours... */
-- 
2.31.1

