--- a/SConstruct	2022-12-09 09:12:36.260399839 +0000
+++ b/SConstruct	2022-12-09 09:15:03.923701343 +0000
@@ -359,7 +359,7 @@
             auto_toolchain_prefix = "armv7l-tizen-linux-gnueabi-"
     elif env['estate'] == '64' and 'v8' in env['arch']:
         if env['os'] == 'linux':
-            auto_toolchain_prefix = "aarch64-linux-gnu-"
+            auto_toolchain_prefix = ""
         elif env['os'] == 'bare_metal':
             auto_toolchain_prefix = "aarch64-elf-"
         elif env['os'] == 'android':
