From 6a8e9ac7f01e92e8a06108753819dbd13eb2da74 Mon Sep 17 00:00:00 2001
From: Max Krummenacher <max.krummenacher@toradex.com>
Date: Wed, 25 Jan 2017 15:08:41 +0100
Subject: [PATCH 2/2] man/Makefile.am: don't error out on missing man
 dependency

Signed-off-by: Max Krummenacher <max.krummenacher@toradex.com>
---
 man/Makefile.am | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/man/Makefile.am b/man/Makefile.am
index fd9fea0..54bb786 100644
--- a/man/Makefile.am
+++ b/man/Makefile.am
@@ -20,9 +20,10 @@ else
 $(man_MANS):
 	@echo ""
 	@echo "Error: you need to run configure with '--enable-man'"
+	@echo "       and the right dependencies installed"
 	@echo "       in order to regenerate man pages."
 	@echo ""
-	@false
+	@echo "man page was not built" > $(man_MANS)
 endif
 
 
-- 
2.5.5

