diff -uprN ros-humble-orocos-kdl-vendor-old/CMakeLists.txt ros-humble-orocos-kdl-vendor/CMakeLists.txt
--- ros-humble-orocos-kdl-vendor-old/CMakeLists.txt	2023-06-12 11:20:15.620878186 +0800
+++ ros-humble-orocos-kdl-vendor/CMakeLists.txt	2023-06-12 11:22:36.452849935 +0800
@@ -13,6 +13,8 @@ if(NOT FORCE_BUILD_VENDOR_PKG)
   find_package(orocos_kdl 1.5.1 QUIET)
 endif()
 
+find_package(orocos_kdl REQUIRED)
+
 macro(build_orocos_kdl)
   set(extra_cmake_args)
   if(DEFINED CMAKE_BUILD_TYPE)
