From 6d1c8fd5830e8f2e967ca9ca5fe45c999116356e Mon Sep 17 00:00:00 2001
From: oee <oee@oee.com>
Date: Tue, 16 Sep 2025 14:41:51 +0800
Subject: [PATCH] fix ot_cipher function err

Signed-off-by: oee <oee@oee.com>


--- a/smp/a55_linux/mpp/component/security_subsys/cipher/v3/src/drv/arch/ss928v100/cipher_cfg.h
+++ b/smp/a55_linux/mpp/component/security_subsys/cipher/v3/src/drv/arch/ss928v100/cipher_cfg.h
@@ -47,7 +47,7 @@
  * #define CRYPTO_SMMU_SUPPORT
  */
 /* support interrupt */
-#define CRYPTO_OS_INT_SUPPORT
+//#define CRYPTO_OS_INT_SUPPORT
 
 /* the hardware version */
 #define CHIP_SYMC_VER_V300
--- a/smp/a55_linux/mpp/component/security_subsys/cipher/v3/src/drv/drivers/core/drv_lib.c
+++ b/smp/a55_linux/mpp/component/security_subsys/cipher/v3/src/drv/drivers/core/drv_lib.c
@@ -253,6 +253,8 @@ td_s32 module_addr_map(td_void)
  */
 td_u32 module_get_secure(td_void)
 {
+    return TD_TRUE;
+    /*
 #ifdef CRYPTO_SEC_CPU
     return TD_TRUE;
 #elif defined(CRYPTO_SWITCH_CPU)
@@ -260,6 +262,7 @@ td_u32 module_get_secure(td_void)
 #else
     return TD_FALSE;
 #endif
+*/
 }
 
 /*
--- a/smp/a55_linux/mpp/component/security_subsys/klad/src/arch/ss928v100/klad_cfg.h
+++ b/smp/a55_linux/mpp/component/security_subsys/klad/src/arch/ss928v100/klad_cfg.h
@@ -35,10 +35,10 @@
 #define KLAD_ROUTE_NUM              2
 
 /* logical module klad support interrupt */
-#define KLAD_INT_SUPPORT
+//#define KLAD_INT_SUPPORT
 
 /* logical module rkp support interrupt */
-#define RKP_INT_SUPPORT
+//#define RKP_INT_SUPPORT
 
 /* interrupt klad number, designated by RKP_INT_RAW.kl_int_num */
 #define KLAD_CLR_ROUTE              0x02
