From 6d1c8fd5830e8f2e967ca9ca5fe45c999116356e Mon Sep 17 00:00:00 2001
From: oee <oee@oee.com>
Date: Tue, 16 Sep 2025 14:41:51 +0800
Subject: [PATCH] fix ot_cipher function err

Signed-off-by: oee <oee@oee.com>

--- a/smp/a55_linux/mpp/component/security_subsys/cipher/v3/src/drv/drivers/core/drv_lib.c
+++ b/smp/a55_linux/mpp/component/security_subsys/cipher/v3/src/drv/drivers/core/drv_lib.c
@@ -253,6 +253,8 @@ td_s32 module_addr_map(td_void)
  */
 td_u32 module_get_secure(td_void)
 {
+    return TD_TRUE;
+    /*
 #ifdef CRYPTO_SEC_CPU
     return TD_TRUE;
 #elif defined(CRYPTO_SWITCH_CPU)
@@ -260,6 +262,7 @@ td_u32 module_get_secure(td_void)
 #else
     return TD_FALSE;
 #endif
+*/
 }
 
 /*
