--- a/sys-cdefs.h	2022-11-16 08:35:20.089145983 +0000
+++ b/sys-cdefs.h	2022-11-16 08:49:14.023211318 +0000
@@ -1,4 +1,5 @@
-#warning usage of non-standard #include <sys/cdefs.h> is deprecated
+#ifndef SYS_CDEFS_H
+#define SYS_CDEFS_H    1
 
 #undef __P
 #undef __PMT
@@ -24,3 +25,14 @@
 # define __THROW
 # define __NTH(fct)     fct
 #endif
+
+#if defined(__GNUC__) && !defined(__cplusplus)
+# define __nonnull(params) __attribute__ ((__nonnull__ params))
+#else
+# define __nonnull(params)
+#endif
+
+#undef __always_inline
+#define __always_inline __inline __attribute__ ((__always_inline__))
+
+#endif   /* sys/cdefs.h */
