From 374555aa3d6ae4c8f054814f1f8088a98f83ec85 Mon Sep 17 00:00:00 2001
From: alichinese <lixinyu44@huawei.com>
Date: Mon, 27 Oct 2025 16:35:41 +0800
Subject: [PATCH] usb.h-Include-sys-types.h

---
 libusb/usb.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/libusb/usb.h b/libusb/usb.h
index ea62543..4f1752d 100644
--- a/libusb/usb.h
+++ b/libusb/usb.h
@@ -51,6 +51,8 @@ typedef unsigned __int32  uint32_t;
 #include <linux/limits.h>
 #endif
 
+#include <sys/types.h>
+
 /*
  * USB spec information
  *
-- 
2.34.1

