From 4758184ce9af69f6ebcd4994fa25d5999f63c63f Mon Sep 17 00:00:00 2001
From: s_c_c <shichuchao@huawei.com>
Date: Tue, 12 Mar 2024 15:08:56 +0800
Subject: [PATCH] feat for embedded comment out unrecognized options

---
 config/compiler/BUILD.gn | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/config/compiler/BUILD.gn b/config/compiler/BUILD.gn
index 0049dc0..508e0ab 100644
--- a/config/compiler/BUILD.gn
+++ b/config/compiler/BUILD.gn
@@ -734,8 +734,8 @@ config("compiler_cpu_abi") {
       }
       cflags += [
         "-march=$arm_arch",
-        "-mfloat-abi=$arm_float_abi",
-        "-mfpu=$arm_fpu",
+        # "-mfloat-abi=$arm_float_abi",
+        # "-mfpu=$arm_fpu",
       ]
       ldflags += [ "-march=$arm_arch" ]
     }
-- 
2.20.1 (Apple Git-117)

