From def2de436600637ff9faa012729b05ef843cd7ad Mon Sep 17 00:00:00 2001
From: s_c_c <shichuchao@huawei.com>
Date: Thu, 5 Sep 2024 19:09:20 +0800
Subject: [PATCH] feat for embedded fix openssl dep

---
 interfaces/innerkits/BUILD.gn | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/interfaces/innerkits/BUILD.gn b/interfaces/innerkits/BUILD.gn
index 65be6f2..19e8160 100644
--- a/interfaces/innerkits/BUILD.gn
+++ b/interfaces/innerkits/BUILD.gn
@@ -49,11 +49,11 @@ ohos_shared_library("libbegetutil") {
     "//base/startup/init/services/param/linux:param_client",
     "//base/startup/init/services/utils:libinit_utils",
   ]
-  deps += [ "//base/startup/init/services/param/base:param_base" ]
+  deps += [ 
+    "//base/startup/init/services/param/base:param_base",
+    "//third_party/openssl:libcrypto_shared",
+    ]
   
-  libs = [
-    "/usr/lib64/libcrypto.so"
-  ]
 
   external_deps = [
     "c_utils:utils",
-- 
2.45.2.windows.1

