diff -uprN git/backend/drm/meson.build git_new/backend/drm/meson.build
--- git/backend/drm/meson.build	2023-10-12 12:56:39.521210488 +0000
+++ git_new/backend/drm/meson.build	2023-10-12 12:57:13.616674278 +0000
@@ -1,10 +1,3 @@
-hwdata = dependency(
-	'hwdata',
-	required: 'drm' in backends,
-	native: true,
-	not_found_message: 'Required for the DRM backend.',
-)
-
 libdisplay_info = dependency(
 	'libdisplay-info',
 	required: 'drm' in backends,
@@ -19,11 +12,11 @@ libliftoff = dependency(
 	required: false,
 )
 
-if not (hwdata.found() and libdisplay_info.found() and features['session'])
+if not (libdisplay_info.found() and features['session'])
 	subdir_done()
 endif
 
-hwdata_dir = hwdata.get_variable(pkgconfig: 'pkgdatadir')
+hwdata_dir = '/usr/share/hwdata'
 pnpids_c = custom_target(
 	'pnpids.c',
 	output: 'pnpids.c',
