diff -Naur ros-humble-slam-toolbox-2.6.4_org/lib/karto_sdk/CMakeLists.txt ros-humble-slam-toolbox-2.6.4/lib/karto_sdk/CMakeLists.txt
--- ros-humble-slam-toolbox-2.6.4_org/lib/karto_sdk/CMakeLists.txt	2023-05-01 00:55:40.985790959 +0800
+++ ros-humble-slam-toolbox-2.6.4/lib/karto_sdk/CMakeLists.txt	2023-05-01 00:56:31.011995373 +0800
@@ -12,7 +12,6 @@
 find_package(rclcpp REQUIRED)
 find_package(Eigen3 REQUIRED)
 find_package(Boost REQUIRED system serialization filesystem thread)
-find_package(TBB REQUIRED NO_CMAKE_PACKAGE_REGISTRY)
 
 set(dependencies
   rclcpp
@@ -29,7 +28,7 @@
 include_directories(include ${EIGEN3_INCLUDE_DIRS} ${TBB_INCLUDE_DIRS} ${Boost_INCLUDE_DIRS})
 add_library(kartoSlamToolbox SHARED src/Karto.cpp src/Mapper.cpp)
 ament_target_dependencies(kartoSlamToolbox ${dependencies})
-target_link_libraries(kartoSlamToolbox ${Boost_LIBRARIES} TBB::tbb)
+target_link_libraries(kartoSlamToolbox ${Boost_LIBRARIES} tbb)
 
 install(DIRECTORY include/
 	DESTINATION include/
