diff -urpN a/frameworks/libhilog/utils/log_utils.cpp b/frameworks/libhilog/utils/log_utils.cpp
--- a/frameworks/libhilog/utils/log_utils.cpp	2024-09-13 08:38:42.265313717 +0000
+++ b/frameworks/libhilog/utils/log_utils.cpp	2024-09-13 08:39:51.054095640 +0000
@@ -494,7 +494,7 @@ void PrintErrorno(int err)
     constexpr int bufSize = 256;
     char buf[bufSize] = { 0 };
 #ifndef __WINDOWS__
-    char *ret = strerror_r(err, buf, bufSize);
+    const char *ret = std::to_string(strerror_r(err, buf, bufSize)).c_str();
     (void)ret;
 #else
     (void)strerror_s(buf, bufSize, err);
