diff -urpN a/frameworks/eventhandler/include/event_handler_utils.h b/frameworks/eventhandler/include/event_handler_utils.h
--- a/frameworks/eventhandler/include/event_handler_utils.h	2024-09-19 02:34:50.626707359 +0000
+++ b/frameworks/eventhandler/include/event_handler_utils.h	2024-09-19 03:02:30.556553452 +0000
@@ -94,7 +94,7 @@ static inline bool AllowHiTraceOutPut(co
 static inline void GetLastErr(char *errmsg, size_t size = MAX_ERRORMSG_LEN)
 {
     size = size > MAX_ERRORMSG_LEN ? MAX_ERRORMSG_LEN : size;
-    char *ret = strerror_r(errno, errmsg, size);
+    const char *ret = std::to_string(strerror_r(errno, errmsg, size)).c_str();
     (void)ret;
 }
 }  // namespace AppExecFwk
