diff -urpN a/toolchain/linux/BUILD.gn b/toolchain/linux/BUILD.gn
--- a/toolchain/linux/BUILD.gn	2024-09-13 06:53:48.291897282 +0000
+++ b/toolchain/linux/BUILD.gn	2024-09-13 06:55:13.602860941 +0000
@@ -150,7 +150,7 @@ clang_toolchain("linux_clang_arm64") {
 
 gcc_toolchain("linux_arm64") {
   tool_path = rebase_path("//toolchain/bin", root_build_dir)
-  toolprefix = "${tool_path}/aarch64-openeuler-linux-gnu-"
+  toolprefix = "${tool_path}/aarch64-openeuler-linux-musl-"
 
   cc = "${toolprefix}gcc"
   cxx = "${toolprefix}g++"
